esphome/config_validation.py

```
@@ -599,9 +599,16 @@ def validator(value):
 
         if isinstance(value, Lambda):
             return returning_lambda(value)
-        if isinstance(other_validators, dict):
+        # If the supplied value does not validate, but is a string like
+        # some_id.state
+        # then turn it into a lambda
+        try:
             return schema(value)
-        return schema(value)
+        except Exception as exc:
+            if isinstance(value, str) and value.endswith(".state"):
```

Since this wont work on strings given the value would be a valid string, we should check that `other_validators` are indeed one of a specific set of simple validators - boolean/float_/uintX/etc

---

Not so simple, the schema can be quite complex (I looked at a few examples.) Right now there is no way I can see of automatically distinguishing between "my_id.state" as a valid string, and `my_id.state` intended to be the state of a text sensor. At this point in the proceedings the id may not have been defined. So it won't work for strings and I am not sure it can be made to, but the standard lambda syntax will still be available.

For bools and floats there is no problem because any string ending with .state will not parse as one of those, so if the validator fails and it ends in `.state` then it can be treated as a lambda. Id validation happens later in the same way as for a regular lambda.

---

Yes, this will only work when the validator on a component is "simple", like `cv.templatable(cv.float_)` this would not support using it on any validator that is more complex than that and should raise a validation error

---

The PR as it is does work with arbitrary validators, except those accepting any string. The usual validation errors are raised. Trying to analyse the validators would not work in all cases, catching the exception does.